### PR TITLE
[DependencyInjection] replace DefinitionDecorator by ChildDefinition

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | TODO

The DefinitionDecorator class does not deal with decorated services. It reflects a parent-child-relationship between definitions instead. To avoid confusion, this commit deprecates the existing DefinitionDecorator class and introduces a new DefinitionExtension class as replacement.

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/SetAclCommandTest.php [hunk 1]
```diff
@@ -54,6 +54,9 @@ protected function tearDown()
         $this->deleteTmpDir('Acl');
     }
 
+    /**
+     * @group legacy
```
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/SetAclCommandTest.php]: This is necessary until DoctrineBundle is compatible with the changes done in this PR.

**Comment by unkind** [comment]
> The DefinitionDecorator class does not deal with decorated services

It reflects service that decorates another one, doesn't it?

**Comment by stof** [comment]
@unkind no. The decorates feature uses normal definition to configure the composition (as the service definition for the decorator does not need to reuse anything from the original service as it gets it as a dependency instead).
DefinitionDecorator is used for the ``parent`` feature, i.e. a definition extending another one

**Comment by nicolas-grekas** [comment]
DerivedDefinition? ChildDefinition?

**Comment by ogizanagi** [comment]
I like `ChildDefinition` 👍

**Comment by julienfalque** [comment]
👍 for `ChildDefinition`.

**Comment by xabbuh** [comment]
I renamed `DefinitionExtension` to `ChildDefinition`.

**Comment by nicolas-grekas** [comment]
I'm tempted to vote +1, but then I also considered the burden that we'll but on bundles & co. Not sure its worth it...

**Comment by xabbuh** [comment]
Given the confusion this class name caused in the past and still causes, I think we should dare to make this deprecation now and get rid of it in 4.0.

**Comment by fabpot** [comment]
👍  There are some conflicts to be resolved.

**Comment by xabbuh** [comment]
conflicts are resolved

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**